### PR TITLE
Signup

### PR DESCRIPTION
**Pull request by cramallo**
## Summary

Implementations(SignUp Trainee): Middleware to validate fields (using express-validator), Service to encrypt passwords (using bcrypt), userController and userService
to create the user indeed;  and errors handling.

## Trello Card

https://trello.com/c/4qyDydh2/2-sign-up

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const { validationResult } = require('express-validator/check');
+const bcrypt = require('bcryptjs');
+const salt = bcrypt.genSaltSync(10);
+const { databaseError } = require('../errors');
+const { bad_request_error } = require('../errors');
+const { conflict_error } = require('../errors');
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: This requires could be merged in one, `const { error1, error2, ... } = require('errors')`

File: app/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+const { validationResult } = require('express-validator/check');
+const bcrypt = require('bcryptjs');
+const salt = bcrypt.genSaltSync(10);
+const { databaseError } = require('../errors');
+const { bad_request_error } = require('../errors');
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Use camel case for errors functions too

File: app/controllers/album.js [hunk 3]
```diff
@@ -11,11 +11,7 @@ exports.getPhotosAlbum = (req, res, next) => {
   return albumService
     .getPhotosAlbum(id)
     .then(albums => {
-      if (albums) {
-        res.send(albums);
-      } else {
-        next(new Error('NOT FOUND'));
-      }
+      res.send(albums);
```
Comment (mpicollo-wolox) [inline app/controllers/album.js]: This change is in multiples PRs

File: app/controllers/user.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+const { conflict_error } = require('../errors');
+const userModel = require('../models').user;
+
+exports.register = (req, res, next) => {
+  const errors = validationResult(req);
+  if (errors.isEmpty() === false) {
+    console.log(errors);
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Remove this `console.log`

File: app/controllers/user.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            .then(userCreated => {
+              console.log(`User with name ${userCreated.name} created!!`);
+              res.send(JSON.stringify(userCreated));
+            })
+            .catch(error => {
+              next(databaseError(error));
+            });
+        }
+      })
+      .catch(error => next(databaseError(error)));
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Specific errors should be handle by services

File: app/controllers/user.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+              lastName: req.body.lastName,
+              email: req.body.email,
+              password: bcrypt.hashSync(req.body.password, salt)
+            })
+            .then(userCreated => {
+              console.log(`User with name ${userCreated.name} created!!`);
+              res.send(JSON.stringify(userCreated));
+            })
+            .catch(error => {
+              next(databaseError(error));
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Specific errors should be handle by services

File: app/controllers/user.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        } else {
+          userModel
+            .create({
+              name: req.body.name,
+              lastName: req.body.lastName,
+              email: req.body.email,
+              password: bcrypt.hashSync(req.body.password, salt)
+            })
+            .then(userCreated => {
+              console.log(`User with name ${userCreated.name} created!!`);
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: We have a logger for this, `console.log` should be avoided

File: app/controllers/user.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+const { bad_request_error } = require('../errors');
+const { conflict_error } = require('../errors');
+const userModel = require('../models').user;
+
+exports.register = (req, res, next) => {
+  const errors = validationResult(req);
+  if (errors.isEmpty() === false) {
+    console.log(errors);
+    next(bad_request_error(errors.array()));
+  } else {
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Avoid nested if-else when is possible, change for this:
```
if (!errors.isEmpty()){
return next(error);
}
... rest of code

```

File: app/controllers/user.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    userModel
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Add return here

File: app/controllers/user.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      .findOne({ where: { email: req.body.email } })
+      .then(user => {
+        if (user) {
+          next(conflict_error('user already exists!'));
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: 409 status code seems don't apply in this case `https://httpstatuses.com/409`

File: app/controllers/user.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        } else {
+          userModel
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Add return here

File: app/controllers/user.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+const { validationResult } = require('express-validator/check');
+const bcrypt = require('bcryptjs');
+const salt = bcrypt.genSaltSync(10);
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Encryption should be handled by a service or utils, also 10 should be a constant or environment variable

File: app/controllers/user.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+          userModel
+            .create({
+              name: req.body.name,
+              lastName: req.body.lastName,
+              email: req.body.email,
+              password: bcrypt.hashSync(req.body.password, salt)
+            })
+            .then(userCreated => {
+              console.log(`User with name ${userCreated.name} created!!`);
+              res.send(JSON.stringify(userCreated));
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: Response should be a JSON not a string

File: app/services/album.js [hunk 14]
```diff
@@ -1,22 +1,15 @@
...
-  new Promise((resolve, reject) => {
-    rp(`${url}albums/${id}/photos`)
-      .then(res => resolve(JSON.parse(res)))
-      .catch(err => {
-        reject(err);
-      });
-  });
+  rp(`${albums_api}albums/${id}/photos`)
+    .then(res => JSON.parse(res))
+    .catch(err => Promise.reject(defaultError(err.message)));
```
Comment (mpicollo-wolox) [inline app/services/album.js]: This not should be updated here

File: app/middlewares/user.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    .isEmpty(),
+  body('lastName', 'lastName error')
+    .not()
+    .isEmpty(),
+  body('email', 'email error')
+    .isEmail()
+    .custom(value => {
+      let aux = value.substr(value.indexOf('@') + 1);
+      aux = aux.substr(0, aux.indexOf('.'));
+      return aux === 'wolox';
```
Comment (mpicollo-wolox) [inline app/middlewares/user.js]: Seems tricky for a simple domain validation, also does not verify `.com.ar`

File: app/controllers/user.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
+const { validationResult } = require('express-validator/check');
+const bcrypt = require('bcryptjs');
+const salt = bcrypt.genSaltSync(10);
+const { databaseError } = require('../errors');
+const { bad_request_error } = require('../errors');
+const { conflict_error } = require('../errors');
+const userModel = require('../models').user;
+
+exports.register = (req, res, next) => {
+  const errors = validationResult(req);
```
Comment (mpicollo-wolox) [inline app/controllers/user.js]: (Nice to have) Could be moved to a middleware

File: app/controllers/user.js (hunk 10 above)
Comment (cramallo) [reply app/controllers/user.js]: As I saw in different forums the 409 status code is the most accurate to use in this case. I could use 400 status code too.
https://stackoverflow.com/questions/1434315/http-status-code-for-database-is-down

Comment (mpicollo-wolox) [reply app/controllers/user.js]: Nice

File: app/controllers/user.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+const userService = require('../services/user');
+
+exports.register = (req, res, next) => {
+  const user = {
+    email: req.body.email,
+    name: req.body.email,
+    lastName: req.body.lastName,
+    password: encryptionService.encryptPassword(req.body.password)
+  };
+  userService
```
Comment (picollomartin) [inline app/controllers/user.js]: Add return here

**Comment by mpicollo-wolox** [comment]
CI fail, check that and request a review later

**Pull request merged**